torchtune/modules/model_fusion/_fusion.py

```diff
@@ -463,18 +500,252 @@ def forward(
+        >>> model = DeepFusionModel(decoder, encoders)
+        >>>
+        >>> # Load full fused checkpoints
+        >>> model.load_state_dict(...)
+        >>>
+        >>> # Or load pretrained individual models (fusion_params are not loaded)
+        >>> model.encoder.load_state_dict(..., strict=False)
+        >>> model.decoder.load_state_dict(..., strict=False)
+        >>>
+        >>> # Forward pass
+        >>> encoder_input = {"image": {...}}
+        >>> output = model(tokens, mask, encoder_input, encoder_mask, input_pos)
+
+    Args:
+        decoder (TransformerDecoder): decoder module
+        encoders (Dict[str, nn.Module]): dictionary mapping encoder name as a string to the encoder module.
+        encoder_tokens (Dict[str, int]): dictionary mapping encoder name to special token ID indicating where
+            in the text sequence the encoder embedding outputs should be injected.
+        decoder_trainable (bool): whether to train or freeze the decoder. Default is False.
+        encoders_trainable (Union[bool, Dict[str, bool]]): whether to train or freeze the encoder. Use a single
+            boolean to set trainable for all encoders or a dictionary keyed by encoder names to specify trainable
+            for each encoder individually. Encoder names should match with ``encoders``. Default is False.
+        fusion_trainable (bool): whether to train the fusion parameters. Default is True.
+
+    Raises:
+        ValueError: if ``encoders`` and ``encoders_trainable`` keys do not match
+    """
+
+    def __init__(
+        self,
+        decoder: TransformerDecoder,
+        encoders: Dict[str, nn.Module],
+        encoder_tokens: Dict[str, int],
+        decoder_trainable: bool = False,
+        encoders_trainable: Union[bool, Dict[str, bool]] = False,
+        fusion_trainable: bool = True,
+    ):
+        super().__init__()
+        if encoders.keys() != encoder_tokens.keys() or (
+            not isinstance(encoders_trainable, bool)
+            and encoders.keys() != encoders_trainable.keys()
+        ):
+            raise ValueError(
+                f"Found mismatched keys in encoders and encoders_trainable. Got {encoders.keys()} and {encoders_trainable.keys()}."
+            )
+
+        self.decoder = decoder
+        self.encoders = nn.ModuleDict(encoders)
+        self.encoder_tokens = encoder_tokens
+        self.encoders_trainable = (
+            {k: encoders_trainable for k in self.encoders.keys()}
+            if isinstance(encoders_trainable, bool)
+            else encoders_trainable
+        )
+
+        # A little surgery in the decoder to give the
+        # fusion module access to control the embeddings
+        # The alternative is to pass a special tok_embeddings
+        # module into TransformerDecoder builder that does the
+        # merging there
+        self.tok_embeddings = decoder.tok_embeddings
+        decoder.tok_embeddings = nn.Identity()
```

Comment:
> If a user wants to use the decoder outside of the EarlyFusionModule in the same script, they will need to restore the original tok_embeddings module from nn.Identity

Why is this a concern? we only set `decoder.tok_embeddings` to identity within EarlyFusionModule? If outside EarlyFusionModule, we just use the normal decoder right?

Reply:
right, but if I passed in the decoder into EarlyFusionModule and modified a layer here, even if I use it separately the embedding layer will have been modified